src/main/java/domain/Janggi.java

```diff
@@ -0,0 +1,164 @@
+import domain.unit.Unit;
+import domain.unit.UnitType;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class Janggi {
+
+    public static final String EMPTY_POINT_EXCEPTION = "해당 위치에 기물이 존재하지 않습니다.";
+    public static final String PICK_OPPOSITE_UNIT_EXCEPTION = "상대팀 말은 고를 수 없습니다.";
+    public static final String CANNOT_MOVE_EXCEPTION = "이동할 수 없는 도착지입니다.";
+
+    private final Map<Position, Unit> units;
+    private Team turn;
+
+    private Janggi(Map<Position, Unit> units, Team turn) {
+        this.units = new HashMap<>(units);
+        this.turn = turn;
+    }
+
+    public static Janggi of(Map<Position, Unit> hanUnits, Map<Position, Unit> choUnits, Team turn) {
+        Map<Position, Unit> units = new HashMap<>();
+        units.putAll(hanUnits);
+        units.putAll(choUnits);
+        return new Janggi(units, turn);
+    }
+
+    public void doTurn(Position pick, Position destination) {
+        if (!canMove(pick, destination)) {
+            throw new IllegalArgumentException(CANNOT_MOVE_EXCEPTION);
+        }
+        Unit pickedUnit = units.get(pick);
+        Unit destinationUnit = units.get(destination);
+        if (destinationUnit != null && destinationUnit.getTeam() == turn.getOpposite()) {
+            units.remove(destination);
+        }
+        units.remove(pick);
+        units.put(destination, pickedUnit);
+        switchTurn();
+    }
+
+    private boolean canMove(Position pick, Position destination) {
+        List<Route> movableRoutes = findMovableRoutesFrom(pick);
+        return movableRoutes.stream()
+                .map(route -> route.searchDestination(pick))
+                .anyMatch(position -> position.equals(destination));
+    }
+
+    private void switchTurn() {
+        turn = turn.getOpposite();
+    }
+
+    public List<Route> findMovableRoutesFrom(Position pick) {
+        if (isEmptyPosition(pick)) {
+            throw new IllegalArgumentException(EMPTY_POINT_EXCEPTION);
+        }
+        Unit pickedUnit = units.get(pick);
+        if (pickedUnit.getTeam() != turn) {
+            throw new IllegalArgumentException(PICK_OPPOSITE_UNIT_EXCEPTION);
+        }
+
+        List<Route> totalRoutes = pickedUnit.calculateRoutes(pick);
+        totalRoutes = filterRoutesByUnitType(pickedUnit, pick, totalRoutes);
+        if (pickedUnit.getType() == UnitType.CANNON) {
+            return totalRoutes;
+        }
+        return filterBlockedRoutes(pick, totalRoutes);
+    }
+
+    private List<Route> filterRoutesByUnitType(Unit pickedUnit, Position pick, List<Route> totalRoutes) {
+        UnitType type = pickedUnit.getType();
+        if (type == UnitType.CANNON) {
+            return totalRoutes.stream()
+                    .filter(route -> canCannonJump(pick, route))
+                    .toList();
+        }
+        if (type == UnitType.SOLDIER) {
+            return filterSoldierMoves(pick, pickedUnit, totalRoutes);
+        }
+        return totalRoutes;
+    }
```

Comment:
한번 할당한 변수를 재할당할 필요 없이 지금처럼 filterRoutesByUnitType 메서드를 수행한 결과를 한번만 변수로 선언하면 될 것 같아요. 😉 

(final은 꼭 안써도되지만, 해당 코멘트를 조금 더 강조하기위해 선언했어요 😃)
```suggestion
        final List<Route> totalRoutes = filterRoutesByUnitType(pickedUnit, pick);
        if (pickedUnit.getType() == UnitType.CANNON) {
            return totalRoutes;
        }
        return filterBlockedRoutes(pick, totalRoutes);
    }

    private List<Route> filterRoutesByUnitType(Unit pickedUnit, Position pick) {
        var routes = pickedUnit.calculateRoutes(pick);
        UnitType type = pickedUnit.getType();
        if (type == UnitType.CANNON) {
            return routes.stream()
                .filter(route -> canCannonJump(pick, route))
                .toList();
        }
        if (type == UnitType.SOLDIER) {
            return filterSoldierMoves(pick, pickedUnit, routes);
        }
        return routes;
    }
```

추가적으로 `filterRoutesByUnitType` 지금처럼 바꾸게되면 행위가 조금 달라지니 네이밍도 바뀌는게 좋을 것 같네요 👀